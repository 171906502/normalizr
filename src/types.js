// @flow
import ArraySchema from './schemas/Array';
import EntitySchema from './schemas/Entity';
import ObjectSchema from './schemas/Object';
import UnionSchema from './schemas/Union';
import ValuesSchema from './schemas/Values';

export type Schema = Array<Schema> | {} | ArraySchema | EntitySchema | ObjectSchema | UnionSchema | ValuesSchema;

export type Visitor = (value: *, parent: ?{}, key: ?string, schema: Schema, addEntity: AddEntity) => *;

export type AddEntity = (schema: EntitySchema, processedEntity: {}, value: {}, parent: *, key: ?string) => void;

export type Unvisitor = (input: *, schema: Schema) => mixed;
