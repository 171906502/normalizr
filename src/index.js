import EntitySchema from './schemas/Entity';
import UnionSchema from './schemas/Union';
import ValuesSchema from './schemas/Values';
import ArraySchema, { normalize as normalizeArray } from './schemas/Array';
import ObjectSchema, { normalize as normalizeObject } from './schemas/Object';

const visit = (value, parent, key, schema, addEntity) => {
  if (typeof value !== 'object' || !value) {
    return value;
  }

  if (typeof schema === 'object' && (!schema.normalize || typeof schema.normalize !== 'function')) {
    let method = normalizeObject;
    if (Array.isArray(schema)) {
      method = normalizeArray;
    }
    return method(schema, value, parent, key, visit, addEntity);
  }

  return schema.normalize(value, parent, key, visit, addEntity);
};

const addEntities = (entities) => (schema, processedEntity, value, parent, key) => {
  const schemaKey = schema.getKey(value, parent, key);
  const id = schema.getId(value, parent, key);
  if (!(schemaKey in entities)) {
    entities[schemaKey] = {};
  }

  const existingEntity = entities[schemaKey][id];
  if (existingEntity) {
    entities[schemaKey][id] = schema.merge(existingEntity, processedEntity);
  } else {
    entities[schemaKey][id] = processedEntity;
  }
};

export const schema = {
  Array: ArraySchema,
  Entity: EntitySchema,
  Object: ObjectSchema,
  Union: UnionSchema,
  Values: ValuesSchema
};

export const normalize = (input, schema) => {
  if (!input || typeof input !== 'object') {
    throw new Error(`Unexpected input given to normalize. Expected type to be "object", found "${typeof input}".`);
  }

  const entities = {};
  const addEntity = addEntities(entities);

  const result = visit(input, input, null, schema, addEntity);
  return { entities, result };
};
