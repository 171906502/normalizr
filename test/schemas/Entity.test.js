/* eslint-env jest */
import { normalize, schema } from '../../src';

describe(schema.Entity.name, () => {
  it('must be created with a key name', () => {
    expect(() => new schema.Entity()).toThrow();
  });

  it('key name must be a string', () => {
    expect(() => new schema.Entity(42)).toThrow();
  });

  it('normalizes an entity', () => {
    const entity = new schema.Entity('item');
    expect(normalize({ id: 1 }, entity)).toMatchSnapshot();
  });

  describe('idAttribute', () => {
    it('can use a custom idAttribute string', () => {
      const user = new schema.Entity('users', {}, { idAttribute: 'id_str' });
      expect(normalize({ id_str: '134351', name: 'Kathy' }, user)).toMatchSnapshot();
    });

    it('can normalize entity IDs based on their object key', () => {
      const user = new schema.Entity('users', {}, { idAttribute: (entity, parent, key) => key });
      const inputSchema = new schema.Values({ users: user }, () => 'users');

      expect(normalize({ 4: { name: 'taco' }, 56: { name: 'burrito' } }, inputSchema)).toMatchSnapshot();
    });

    it('can build the entity\'s ID from the parent object', () => {
      const user = new schema.Entity('users', {}, {
        idAttribute: (entity, parent, key) => `${parent.name}-${key}-${entity.id}`
      });
      const inputSchema = new schema.Object({ user });

      expect(normalize({ name: 'tacos', user: { id: '4', name: 'Jimmy' } }, inputSchema)).toMatchSnapshot();
    });
  });

  describe('mergeStrategy', () => {
    it('can use a custom merging strategy', () => {
      const mergeStrategy = (entityA, entityB) => {
        return { ...entityA, ...entityB, name: entityA.name };
      };
      const mySchema = new schema.Entity('tacos', {}, { mergeStrategy });
      const inputSchema = new schema.Array(mySchema);

      expect(normalize([
        { id: 1, name: 'foo' },
        { id: 1, name: 'bar', alias: 'bar' }
      ], inputSchema)).toMatchSnapshot();
    });
  });

  describe('processStrategy', () => {
    it('can use a custom processing strategy', () => {
      const processStrategy = (entity) => {
        return { ...entity, slug: `thing-${entity.id}` };
      };
      const mySchema = new schema.Entity('tacos', {}, { processStrategy });

      expect(normalize({ id: 1, name: 'foo' }, mySchema)).toMatchSnapshot();
    });

    it('can use information from the parent in the process strategy', () => {
      const processStrategy = (entity, parent, key) => {
        return { ...entity, parentId: parent.id, parentKey: key };
      };
      const childEntity = new schema.Entity('children', {}, { processStrategy });
      const parentEntity = new schema.Entity('parents', {
        child: childEntity
      });

      expect(normalize({
        id: 1, content: 'parent', childKey: { id: 4, content: 'child' }
      }, parentEntity)).toMatchSnapshot();
    });
  });
});
