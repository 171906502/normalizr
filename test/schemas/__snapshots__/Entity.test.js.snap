exports[`EntitySchema can build the entity's ID from the parent object 1`] = `
Object {
  "entities": Object {
    "users": Object {
      "tacos-4": Object {
        "id": "4",
        "name": "Jimmy",
      },
    },
  },
  "result": Object {
    "name": "tacos",
    "user": "tacos-4",
  },
}
`;

exports[`EntitySchema can normalize entity IDs based on their object key 1`] = `
Object {
  "entities": Object {
    "users": Object {
      "4": Object {
        "name": "taco",
      },
      "56": Object {
        "name": "burrito",
      },
    },
  },
  "result": Object {
    "4": "4",
    "56": "56",
  },
}
`;

exports[`EntitySchema can use a custom merging strategy 1`] = `
Object {
  "entities": Object {
    "tacos": Object {
      "1": Object {
        "alias": "bar",
        "id": 1,
        "name": "foo",
      },
    },
  },
  "result": Array [
    1,
    1,
  ],
}
`;

exports[`EntitySchema can use a custom processing strategy 1`] = `
Object {
  "entities": Object {
    "tacos": Object {
      "1": Object {
        "id": 1,
        "name": "foo",
        "slug": "thing-1",
      },
    },
  },
  "result": 1,
}
`;

exports[`EntitySchema normalizes an entity 1`] = `
Object {
  "entities": Object {
    "item": Object {
      "1": Object {
        "id": 1,
      },
    },
  },
  "result": 1,
}
`;
